src/cpp/src/llm_pipeline_stateful.cpp

```diff
@@ -211,17 +211,23 @@ EncodedResults StatefulLLMPipeline::generate(
     if (!is_chat_conversation) {
         reset_kv_state();
         m_model_runner.get_tensor("attention_mask").set_shape({1, 0});
+        m_kv_cache_state.reset_state();
+        m_kv_history_trim_manager.reset();
     }
```

Comment:
why don't we reset all states at the end of previous `generate()` ?
I this case states (KV caches) could be freed earlier

Reply:
The downside of resetting all states at the end is lack of exception safety. If the exception is thrown in the middle of `generate()`, the cache won't be reset.
Another reason is performance. There's no reason to reset anything if it was the last ever `generate()`.

The third approach would be to reset state in the beginning and the end of `generate()`. But I don't see any strong argument to prefer any of these three approaches.

Reply:
> Another reason is performance. There's no reason to reset anything if it was the last ever generate().

Reset should be a cheap operation and it will happen anyway during pipeline destruction
So, for last even generate() we just move this part to generate instead of ~Pipeline()